diff --git a/src/pgtkfns.c b/src/pgtkfns.c
index 4257dc45f98..41f095982cf 100644
--- a/src/pgtkfns.c
+++ b/src/pgtkfns.c
@@ -54,6 +54,9 @@ Copyright (C) 1989, 1992-1994, 2005-2006, 2008-2020, 2022-2026 Free
 static double
 pgtk_get_monitor_scale_factor (const char *model)
 {
+  if (NILP (monitor_scale_factor_alist))
+    return 0.0;
+
   if (model == NULL)
     return 0.0;
 
@@ -247,7 +250,7 @@ pgtk_set_border_color (struct frame *f, Lisp_Object arg, Lisp_Object oldval)
   CHECK_STRING (arg);
   pix = x_decode_color (f, arg, BLACK_PIX_DEFAULT (f));
   FRAME_X_OUTPUT (f)->border_pixel = pix;
-  pgtk_frame_rehighlight (FRAME_DISPLAY_INFO (f));
+  pgtk_frame_rehighlight_forced (FRAME_DISPLAY_INFO (f));
 }
 
 static void
@@ -791,12 +794,15 @@ xg_set_icon_from_xpm_data (struct frame *f, const char **data)
   if (!pixbuf)
     return false;
 
-  if (!FRAME_GTK_OUTER_WIDGET (f))
-    return false;
+  if (FRAME_GTK_OUTER_WIDGET (f))
+    {
+      gtk_window_set_icon (GTK_WINDOW (FRAME_GTK_OUTER_WIDGET (f)), pixbuf);
+      g_object_unref (pixbuf);
+      return true;
+    }
 
-  gtk_window_set_icon (GTK_WINDOW (FRAME_GTK_OUTER_WIDGET (f)), pixbuf);
   g_object_unref (pixbuf);
-  return true;
+  return false;
 }
 
 static void
@@ -806,6 +812,9 @@ pgtk_set_sticky (struct frame *f, Lisp_Object new_value,
   if (!FRAME_GTK_OUTER_WIDGET (f))
     return;
 
+  if (EQ (new_value, old_value))
+    return;
+
   if (!NILP (new_value))
     gtk_window_stick (GTK_WINDOW (FRAME_GTK_OUTER_WIDGET (f)));
   else
@@ -840,6 +849,9 @@ pgtk_set_scroll_bar_foreground (struct frame *f, Lisp_Object new_value,
   if (FRAME_TOOLTIP_P (f))
     return;
 
+  if (!NILP (Fequal (new_value, old_value)))
+    return;
+
   if (NILP (new_value))
     {
       gtk_css_provider_load_from_data (css_provider, "", -1, NULL);
@@ -870,6 +882,9 @@ pgtk_set_scroll_bar_background (struct frame *f, Lisp_Object new_value,
   GtkCssProvider *css_provider =
     FRAME_X_OUTPUT (f)->scrollbar_background_css_provider;
 
+  if (!NILP (Fequal (new_value, old_value)))
+    return;
+
   if (NILP (new_value))
     {
       gtk_css_provider_load_from_data (css_provider, "", -1, NULL);
@@ -1635,8 +1650,9 @@ #define INSTALL_CURSOR(FIELD, NAME) \
   if (FRAME_GTK_OUTER_WIDGET (f))
     {
       GList *w = gtk_container_get_children (GTK_CONTAINER (FRAME_GTK_OUTER_WIDGET (f)));
-      for (; w != NULL; w = w->next)
-	gtk_widget_show_all (GTK_WIDGET (w->data));
+      for (GList *iter = w; iter != NULL; iter = iter->next)
+	gtk_widget_show_all (GTK_WIDGET (iter->data));
+      g_list_free (w);
     }
 
   gui_default_parameter (f, parms, Qno_focus_on_map, Qnil,
@@ -1737,7 +1753,7 @@ #define INSTALL_CURSOR(FIELD, NAME) \
 
   FRAME_X_OUTPUT (f)->cr_surface_visible_bell = NULL;
   FRAME_X_OUTPUT (f)->atimer_visible_bell = NULL;
-  FRAME_X_OUTPUT (f)->watched_scale_factor = 1.0;
+  FRAME_X_OUTPUT (f)->watched_scale_factor = FRAME_SCALE_FACTOR (f);
   struct timespec ts = make_timespec (1, 0);
   FRAME_X_OUTPUT (f)->scale_factor_atimer = start_atimer(ATIMER_CONTINUOUS,
 							 ts,
@@ -2012,6 +2028,7 @@ DEFUN ("pgtk-set-resource", Fpgtk_set_resource, Spgtk_set_resource, 2, 2, 0,
   char *key = g_strdup_printf ("%s.%s", res, attr);
 
   pgtk_set_defaults_value (key, NILP (value) ? NULL : SSDATA (value));
+  g_free (key);
 
   return Qnil;
 }
@@ -2427,12 +2444,17 @@ DEFUN ("pgtk-display-monitor-attributes-list", Fpgtk_display_monitor_attributes_
   Lisp_Object monitor_frames, rest, frame;
   static const char *source = "Gdk";
   struct MonitorInfo *monitors;
+  GdkMonitor **gdk_monitors;
 
   block_input ();
   gdpy = dpyinfo->gdpy;
   n_monitors = gdk_display_get_n_monitors (gdpy);
   monitor_frames = make_nil_vector (n_monitors);
   monitors = xzalloc (n_monitors * sizeof *monitors);
+  gdk_monitors = xnmalloc (n_monitors, sizeof *gdk_monitors);
+
+  for (i = 0; i < n_monitors; ++i)
+    gdk_monitors[i] = gdk_display_get_monitor (gdpy, i);
 
   FOR_EACH_FRAME (rest, frame)
     {
@@ -2443,11 +2465,12 @@ DEFUN ("pgtk-display-monitor-attributes-list", Fpgtk_display_monitor_attributes_
 	  && !FRAME_TOOLTIP_P (f))
 	{
 	  GdkWindow *gwin = gtk_widget_get_window (FRAME_GTK_WIDGET (f));
+	  GdkMonitor *frame_monitor
+	    = gdk_display_get_monitor_at_window (gdpy, gwin);
 
-          for (i = 0; i < n_monitors; i++)
-            if (gdk_display_get_monitor_at_window (gdpy, gwin)
-                == gdk_display_get_monitor (gdpy, i))
-              break;
+	  for (i = 0; i < n_monitors; i++)
+	    if (frame_monitor == gdk_monitors[i])
+	      break;
 	  ASET (monitor_frames, i, Fcons (frame, AREF (monitor_frames, i)));
 	}
     }
@@ -2459,7 +2482,7 @@ DEFUN ("pgtk-display-monitor-attributes-list", Fpgtk_display_monitor_attributes_
       struct MonitorInfo *mi = &monitors[i];
       double scale = 1;
 
-      GdkMonitor *monitor = gdk_display_get_monitor (gdpy, i);
+      GdkMonitor *monitor = gdk_monitors[i];
       if (gdk_monitor_is_primary (monitor))
         primary_monitor = i;
       gdk_monitor_get_geometry (monitor, &rec);
@@ -2502,6 +2525,7 @@ DEFUN ("pgtk-display-monitor-attributes-list", Fpgtk_display_monitor_attributes_
                                                  monitor_frames,
                                                  source);
   free_monitors (monitors, n_monitors);
+  xfree (gdk_monitors);
   unblock_input ();
 
   return attributes_list;
@@ -2860,41 +2884,36 @@ compute_tip_xy (struct frame *f, Lisp_Object parms, Lisp_Object dx,
   if ((!FIXNUMP (left) && !FIXNUMP (right))
       || (!FIXNUMP (top) && !FIXNUMP (bottom)))
     {
-      Lisp_Object frame, attributes, monitor, geometry;
+      GdkDisplay *gdpy = FRAME_DISPLAY_INFO (f)->gdpy;
       GdkSeat *seat =
-	gdk_display_get_default_seat (FRAME_DISPLAY_INFO (f)->gdpy);
+	gdk_display_get_default_seat (gdpy);
       GdkDevice *dev = gdk_seat_get_pointer (seat);
+      GdkMonitor *monitor;
       GdkScreen *scr;
 
       block_input ();
       gdk_device_get_position (dev, &scr, root_x, root_y);
-      unblock_input ();
-
-      XSETFRAME (frame, f);
-      attributes = Fpgtk_display_monitor_attributes_list (frame);
+      monitor = gdk_display_get_monitor_at_point (gdpy, *root_x, *root_y);
 
       /* Try to determine the monitor where the mouse pointer is and
          its geometry.  See bug#22549.  */
-      while (CONSP (attributes))
+      if (monitor)
 	{
-	  monitor = XCAR (attributes);
-	  geometry = Fassq (Qgeometry, monitor);
-	  if (CONSP (geometry))
-	    {
-	      min_x = XFIXNUM (Fnth (make_fixnum (1), geometry));
-	      min_y = XFIXNUM (Fnth (make_fixnum (2), geometry));
-	      max_x = min_x + XFIXNUM (Fnth (make_fixnum (3), geometry));
-	      max_y = min_y + XFIXNUM (Fnth (make_fixnum (4), geometry));
-	      if (min_x <= *root_x && *root_x < max_x
-		  && min_y <= *root_y && *root_y < max_y)
-		{
-		  break;
-		}
-	      max_y = -1;
-	    }
-
-	  attributes = XCDR (attributes);
+	  GdkRectangle rec;
+	  double scale;
+
+	  gdk_monitor_get_geometry (monitor, &rec);
+	  scale = pgtk_get_monitor_scale_factor (gdk_monitor_get_model (monitor));
+	  if (scale == 0.0)
+	    scale = gdk_monitor_get_scale_factor (monitor);
+
+	  min_x = rec.x * scale + 0.5;
+	  min_y = rec.y * scale + 0.5;
+	  max_x = min_x + rec.width * scale + 0.5;
+	  max_y = min_y + rec.height * scale + 0.5;
 	}
+
+      unblock_input ();
     }
 
   /* It was not possible to determine the monitor's geometry, so we
diff --git a/src/pgtkim.c b/src/pgtkim.c
index 81d9503cbce..32b81caae0b 100644
--- a/src/pgtkim.c
+++ b/src/pgtkim.c
@@ -68,6 +68,7 @@ im_context_preedit_changed_cb (GtkIMContext *imc, gpointer user_data)
   struct frame *f = dpyinfo->im.focused_frame;
   char *str;
   PangoAttrList *attrs;
+  ptrdiff_t len;
   int pos;
 
   if (dpyinfo->im.context == NULL)
@@ -76,6 +77,7 @@ im_context_preedit_changed_cb (GtkIMContext *imc, gpointer user_data)
     return;
 
   gtk_im_context_get_preedit_string (imc, &str, &attrs, &pos);
+  len = strlen (str);
 
 
   /*
@@ -96,8 +98,8 @@ im_context_preedit_changed_cb (GtkIMContext *imc, gpointer user_data)
 
       pango_attr_iterator_range (iter, &st, &ed);
 
-      if (ed > strlen (str))
-	ed = strlen (str);
+      if (ed > len)
+	ed = len;
       if (st >= ed)
 	continue;
 
@@ -144,6 +146,7 @@ im_context_preedit_changed_cb (GtkIMContext *imc, gpointer user_data)
   list = Fnreverse (list);
   pgtk_enqueue_preedit (f, list);
 
+  pango_attr_iterator_destroy (iter);
   g_free (str);
   pango_attr_list_unref (attrs);
 }
diff --git a/src/pgtkselect.c b/src/pgtkselect.c
index 425da3b8fd4..7fe9799afc3 100644
--- a/src/pgtkselect.c
+++ b/src/pgtkselect.c
@@ -138,7 +138,10 @@ symbol_to_gdk_atom (Lisp_Object sym)
 static Lisp_Object
 gdk_atom_to_symbol (GdkAtom atom)
 {
-  return intern (gdk_atom_name (atom));
+  gchar *name = gdk_atom_name (atom);
+  Lisp_Object symbol = intern (name);
+  g_free (name);
+  return symbol;
 }
 
 
@@ -630,6 +633,8 @@ pgtk_handle_selection_request (struct selection_input_event *event)
 			   cs.type, cs.format,
 			   GDK_PROP_MODE_REPLACE,
 			   cs.data, cs.size);
+      if (!cs.nofree && cs.data)
+	xfree (cs.data);
       success = true;
     }
   else
@@ -1092,7 +1097,8 @@ pgtk_size_for_format (gint format)
 /* Use xfree, not g_free, to free the data obtained with this function.  */
 
 static void
-pgtk_get_window_property (GdkWindow *window, unsigned char **data_ret,
+pgtk_get_window_property (struct pgtk_display_info *dpyinfo,
+			  GdkWindow *window, unsigned char **data_ret,
 			  ptrdiff_t *bytes_ret, GdkAtom *actual_type_ret,
 			  int *actual_format_ret, unsigned long *actual_size_ret)
 {
@@ -1123,7 +1129,7 @@ pgtk_get_window_property (GdkWindow *window, unsigned char **data_ret,
     }
 
   if (actual_type == GDK_SELECTION_TYPE_ATOM
-      || actual_type == gdk_atom_intern_static_string ("ATOM_PAIR"))
+      || actual_type == dpyinfo->atom_ATOM_PAIR)
     {
       /* GDK should not allow anything else.  */
       eassert (actual_format == 32);
@@ -1189,7 +1195,7 @@ pgtk_get_window_property_as_lisp_data (struct pgtk_display_info *dpyinfo,
   Lisp_Object val;
   GdkDisplay *display = dpyinfo->display;
 
-  pgtk_get_window_property (window, &data, &bytes,
+  pgtk_get_window_property (dpyinfo, window, &data, &bytes,
 			    &actual_type, &actual_format,
 			    &actual_size);
 
@@ -1261,7 +1267,7 @@ selection_data_to_lisp_data (struct pgtk_display_info *dpyinfo,
 			     const unsigned char *data,
 			     ptrdiff_t size, GdkAtom type, int format)
 {
-  if (type == gdk_atom_intern_static_string ("NULL"))
+  if (type == dpyinfo->atom_NULL)
     return QNULL;
   /* Convert any 8-bit data to a string, for compactness.  */
   else if (format == 8)
@@ -1273,9 +1279,9 @@ selection_data_to_lisp_data (struct pgtk_display_info *dpyinfo,
 	 property `foreign-selection' so that the caller of
 	 x-get-selection-internal (usually x-get-selection) can know
 	 that the string must be decode.  */
-      if (type == gdk_atom_intern_static_string ("COMPOUND_TEXT"))
+      if (type == dpyinfo->atom_COMPOUND_TEXT)
 	lispy_type = QCOMPOUND_TEXT;
-      else if (type == gdk_atom_intern_static_string ("UTF8_STRING"))
+      else if (type == dpyinfo->atom_UTF8_STRING)
 	lispy_type = QUTF8_STRING;
       else
 	lispy_type = QSTRING;
@@ -1289,7 +1295,7 @@ selection_data_to_lisp_data (struct pgtk_display_info *dpyinfo,
   else if (format == 32
 	   && (type == GDK_SELECTION_TYPE_ATOM
 	       /* Treat ATOM_PAIR type similar to list of atoms.  */
-	       || type == gdk_atom_intern_static_string ("ATOM_PAIR")))
+	       || type == dpyinfo->atom_ATOM_PAIR))
     {
       ptrdiff_t i;
       GdkAtom *idata = (GdkAtom *) data;
diff --git a/src/pgtkterm.c b/src/pgtkterm.c
index 757ff57a9f2..93543c0c9d0 100644
--- a/src/pgtkterm.c
+++ b/src/pgtkterm.c
@@ -84,7 +84,7 @@ #define FRAME_CR_SURFACE(f)		(cairo_get_target (FRAME_CR_CONTEXT (f)))
 struct event_queue_t
 {
   union buffered_input_event *q;
-  int nr, cap;
+  int head, nr, cap;
 };
 
 /* A queue of events that will be read by the read_socket_hook.  */
@@ -231,6 +231,8 @@ pgtk_free_devices (struct pgtk_display_info *dpyinfo)
     }
 
   dpyinfo->devices = NULL;
+  dpyinfo->last_event_device = NULL;
+  dpyinfo->last_event_device_name = Qnil;
 }
 
 static void
@@ -252,12 +254,21 @@ pgtk_get_device_for_event (struct pgtk_display_info *dpyinfo,
   if (!device)
     return Qt;
 
+  if (device == dpyinfo->last_event_device)
+    return dpyinfo->last_event_device_name;
+
   for (tem = dpyinfo->devices; tem; tem = tem->next)
     {
       if (tem->device == device)
-	return tem->name;
+	{
+	  dpyinfo->last_event_device = device;
+	  dpyinfo->last_event_device_name = tem->name;
+	  return tem->name;
+	}
     }
 
+  dpyinfo->last_event_device = device;
+  dpyinfo->last_event_device_name = Qt;
   return Qt;
 }
 
@@ -290,19 +301,32 @@ evq_enqueue (union buffered_input_event *ev)
   struct frame *frame;
   struct pgtk_display_info *dpyinfo;
 
-  if (evq->cap == 0)
-    {
-      evq->cap = 4;
-      evq->q = xmalloc (sizeof *evq->q * evq->cap);
-    }
-
   if (evq->nr >= evq->cap)
     {
-      evq->cap += evq->cap / 2;
-      evq->q = xrealloc (evq->q, sizeof *evq->q * evq->cap);
+      int oldcap = evq->cap;
+      int newcap = oldcap ? oldcap + oldcap / 2 : 4;
+      union buffered_input_event *newq;
+      int idx = evq->head;
+
+      newq = xmalloc (sizeof *newq * newcap);
+      for (int i = 0; i < evq->nr; i++)
+	{
+	  newq[i] = evq->q[idx];
+	  if (++idx == oldcap)
+	    idx = 0;
+	}
+
+      xfree (evq->q);
+      evq->q = newq;
+      evq->head = 0;
+      evq->cap = newcap;
     }
 
-  evq->q[evq->nr++] = *ev;
+  int tail = evq->head + evq->nr;
+  if (tail >= evq->cap)
+    tail -= evq->cap;
+  evq->q[tail] = *ev;
+  evq->nr++;
 
   if (ev->ie.kind != SELECTION_REQUEST_EVENT
       && ev->ie.kind != SELECTION_CLEAR_EVENT)
@@ -339,11 +363,12 @@ evq_flush (struct input_event *hold_quit)
 	 we need to shift event queue first and pass the event
 	 to kbd_buffer_store_buffered_event so that events in
 	 queue are not processed twice.  Bug#52941 */
-      union buffered_input_event ev = evq->q[0];
-      int i;
-      for (i = 1; i < evq->nr; i++)
-	evq->q[i - 1] = evq->q[i];
+      union buffered_input_event ev = evq->q[evq->head];
+      if (++evq->head == evq->cap)
+	evq->head = 0;
       evq->nr--;
+      if (evq->nr == 0)
+	evq->head = 0;
 
       kbd_buffer_store_buffered_event (&ev, hold_quit);
       n++;
@@ -359,10 +384,13 @@ mark_pgtkterm (void)
   struct pgtk_device_t *device;
   struct event_queue_t *evq = &event_q;
   int i, n = evq->nr;
+  int idx = evq->head;
 
   for (i = 0; i < n; i++)
     {
-      union buffered_input_event *ev = &evq->q[i];
+      union buffered_input_event *ev = &evq->q[idx];
+      if (++idx == evq->cap)
+	idx = 0;
 
       /* Selection requests don't have Lisp object members.  */
 
@@ -383,6 +411,8 @@ mark_pgtkterm (void)
       for (device = dpyinfo->devices; device;
 	   device = device->next)
 	mark_object (device->name);
+      if (dpyinfo->last_event_device)
+	mark_object (dpyinfo->last_event_device_name);
     }
 }
 
@@ -565,31 +595,32 @@ pgtk_calc_absolute_position (struct frame *f)
 {
   int flags = f->size_hint_flags;
   struct frame *p = FRAME_PARENT_FRAME (f);
+  Lisp_Object outer_edges = Qnil;
 
   /* We have nothing to do if the current position
      is already for the top-left corner.  */
   if (!((flags & XNegative) || (flags & YNegative)))
     return;
 
+  /* A frame that has been visible at least once should have outer
+     edges.  */
+  if (f->output_data.pgtk->has_been_visible && !p)
+    {
+      Lisp_Object frame;
+
+      XSETFRAME (frame, f);
+      outer_edges = Fpgtk_frame_edges (frame, Qouter_edges);
+    }
+
   /* Treat negative positions as relative to the leftmost bottommost
      position that fits on the screen.  */
   if (flags & XNegative)
     {
       int width = FRAME_PIXEL_WIDTH (f);
 
-      /* A frame that has been visible at least once should have outer
-	 edges.  */
-      if (f->output_data.pgtk->has_been_visible && !p)
-	{
-	  Lisp_Object frame;
-	  Lisp_Object edges = Qnil;
-
-	  XSETFRAME (frame, f);
-	  edges = Fpgtk_frame_edges (frame, Qouter_edges);
-	  if (!NILP (edges))
-	    width = (XFIXNUM (Fnth (make_fixnum (2), edges))
-		     - XFIXNUM (Fnth (make_fixnum (0), edges)));
-	}
+      if (!NILP (outer_edges))
+	width = (XFIXNUM (Fnth (make_fixnum (2), outer_edges))
+		 - XFIXNUM (Fnth (make_fixnum (0), outer_edges)));
 
       if (p)
 	f->left_pos = (FRAME_PIXEL_WIDTH (p) - width - 2 * f->border_width
@@ -604,18 +635,9 @@ pgtk_calc_absolute_position (struct frame *f)
     {
       int height = FRAME_PIXEL_HEIGHT (f);
 
-      if (f->output_data.pgtk->has_been_visible && !p)
-	{
-	  Lisp_Object frame;
-	  Lisp_Object edges = Qnil;
-
-	  XSETFRAME (frame, f);
-	  if (NILP (edges))
-	    edges = Fpgtk_frame_edges (frame, Qouter_edges);
-	  if (!NILP (edges))
-	    height = (XFIXNUM (Fnth (make_fixnum (3), edges))
-		      - XFIXNUM (Fnth (make_fixnum (1), edges)));
-	}
+      if (!NILP (outer_edges))
+	height = (XFIXNUM (Fnth (make_fixnum (3), outer_edges))
+		  - XFIXNUM (Fnth (make_fixnum (1), outer_edges)));
 
       if (p)
 	f->top_pos = (FRAME_PIXEL_HEIGHT (p) - height - 2 * f->border_width
@@ -1026,6 +1048,9 @@ pgtk_set_no_focus_on_map (struct frame *f, Lisp_Object new_value,
 pgtk_set_no_accept_focus (struct frame *f, Lisp_Object new_value,
 			  Lisp_Object old_value)
 {
+  if (EQ (new_value, old_value))
+    return;
+
   xg_set_no_accept_focus (f, new_value);
   FRAME_NO_ACCEPT_FOCUS (f) = !NILP (new_value);
 }
@@ -1036,6 +1061,9 @@ pgtk_set_z_group (struct frame *f, Lisp_Object new_value, Lisp_Object old_value)
   if (!FRAME_GTK_OUTER_WIDGET (f))
     return;
 
+  if (EQ (new_value, old_value))
+    return;
+
   if (NILP (new_value))
     {
       gtk_window_set_keep_above (GTK_WINDOW (FRAME_GTK_OUTER_WIDGET (f)),
@@ -1085,6 +1113,10 @@ pgtk_initialize_display_info (struct pgtk_display_info *dpyinfo)
   dpyinfo->n_fonts = 0;
   dpyinfo->smallest_font_height = 1;
   dpyinfo->smallest_char_width = 1;
+  dpyinfo->atom_ATOM_PAIR = gdk_atom_intern_static_string ("ATOM_PAIR");
+  dpyinfo->atom_COMPOUND_TEXT = gdk_atom_intern_static_string ("COMPOUND_TEXT");
+  dpyinfo->atom_NULL = gdk_atom_intern_static_string ("NULL");
+  dpyinfo->atom_UTF8_STRING = gdk_atom_intern_static_string ("UTF8_STRING");
 
   reset_mouse_highlight (&dpyinfo->mouse_highlight);
 }
@@ -1297,12 +1329,12 @@ pgtk_clear_glyph_string_rect (struct glyph_string *s, int x, int y,
 }
 
 static void
-fill_background_by_face (struct frame *f, struct face *face, int x, int y,
-			 int width, int height)
+fill_background_by_face_1 (struct frame *f, struct face *face, cairo_t *cr,
+			   int x, int y, int width, int height)
 {
-  cairo_t *cr = pgtk_begin_cr_clip (f);
   double r, g, b, a;
 
+  cairo_save (cr);
   cairo_set_operator (cr, CAIRO_OPERATOR_SOURCE);
   cairo_rectangle (cr, x, y, width, height);
   cairo_clip (cr);
@@ -1326,6 +1358,16 @@ fill_background_by_face (struct frame *f, struct face *face, int x, int y,
       cairo_mask (cr, mask);
     }
 
+  cairo_restore (cr);
+}
+
+static void
+fill_background_by_face (struct frame *f, struct face *face, int x, int y,
+			 int width, int height)
+{
+  cairo_t *cr = pgtk_begin_cr_clip (f);
+
+  fill_background_by_face_1 (f, face, cr, x, y, width, height);
   pgtk_end_cr_clip (f);
 }
 
@@ -1853,8 +1895,9 @@ pgtk_draw_relief_rect (struct frame *f,
 
   if (left_p)
     {
-      pgtk_fill_rectangle (f, top_left_color, left_x, top_y,
-			   vwidth, bottom_y + 1 - top_y, false);
+      pgtk_set_cr_source_with_color (f, top_left_color, false);
+      cairo_rectangle (cr, left_x, top_y, vwidth, bottom_y + 1 - top_y);
+      cairo_fill (cr);
       if (top_p)
 	corners |= 1 << CORNER_TOP_LEFT;
       if (bot_p)
@@ -1862,8 +1905,10 @@ pgtk_draw_relief_rect (struct frame *f,
     }
   if (right_p)
     {
-      pgtk_fill_rectangle (f, bottom_right_color, right_x + 1 - vwidth, top_y,
-			   vwidth, bottom_y + 1 - top_y, false);
+      pgtk_set_cr_source_with_color (f, bottom_right_color, false);
+      cairo_rectangle (cr, right_x + 1 - vwidth, top_y,
+		       vwidth, bottom_y + 1 - top_y);
+      cairo_fill (cr);
       if (top_p)
 	corners |= 1 << CORNER_TOP_RIGHT;
       if (bot_p)
@@ -1872,8 +1917,11 @@ pgtk_draw_relief_rect (struct frame *f,
   if (top_p)
     {
       if (!right_p)
-	pgtk_fill_rectangle (f, top_left_color, left_x, top_y,
-			     right_x + 1 - left_x, hwidth, false);
+	{
+	  pgtk_set_cr_source_with_color (f, top_left_color, false);
+	  cairo_rectangle (cr, left_x, top_y, right_x + 1 - left_x, hwidth);
+	  cairo_fill (cr);
+	}
       else
 	pgtk_fill_trapezoid_for_relief (f, top_left_color, left_x, top_y,
 					right_x + 1 - left_x, hwidth, 1);
@@ -1881,20 +1929,29 @@ pgtk_draw_relief_rect (struct frame *f,
   if (bot_p)
     {
       if (!left_p)
-	pgtk_fill_rectangle (f, bottom_right_color, left_x,
-			     bottom_y + 1 - hwidth, right_x + 1 - left_x,
-			     hwidth, false);
+	{
+	  pgtk_set_cr_source_with_color (f, bottom_right_color, false);
+	  cairo_rectangle (cr, left_x, bottom_y + 1 - hwidth,
+			   right_x + 1 - left_x, hwidth);
+	  cairo_fill (cr);
+	}
       else
 	pgtk_fill_trapezoid_for_relief (f, bottom_right_color,
 					left_x, bottom_y + 1 - hwidth,
 					right_x + 1 - left_x, hwidth, 0);
     }
   if (left_p && vwidth > 1)
-    pgtk_fill_rectangle (f, bottom_right_color, left_x, top_y,
-			 1, bottom_y + 1 - top_y, false);
+    {
+      pgtk_set_cr_source_with_color (f, bottom_right_color, false);
+      cairo_rectangle (cr, left_x, top_y, 1, bottom_y + 1 - top_y);
+      cairo_fill (cr);
+    }
   if (top_p && hwidth > 1)
-    pgtk_fill_rectangle (f, bottom_right_color, left_x, top_y,
-			 right_x + 1 - left_x, 1, false);
+    {
+      pgtk_set_cr_source_with_color (f, bottom_right_color, false);
+      cairo_rectangle (cr, left_x, top_y, right_x + 1 - left_x, 1);
+      cairo_fill (cr);
+    }
   if (corners)
     pgtk_erase_corners_for_relief (f, FRAME_BACKGROUND_PIXEL (f), left_x,
 				   top_y, right_x - left_x + 1,
@@ -1916,39 +1973,28 @@ pgtk_draw_box_rect (struct glyph_string *s, int left_x,
 		    int vwidth, bool left_p, bool right_p,
 		    XRectangle * clip_rect)
 {
-  unsigned long foreground_backup;
-
   cairo_t *cr = pgtk_begin_cr_clip (s->f);
 
-  foreground_backup = s->xgcv.foreground;
-  s->xgcv.foreground = s->face->box_color;
-
   pgtk_set_clip_rectangles (s->f, cr, clip_rect, 1);
+  pgtk_set_cr_source_with_color (s->f, s->face->box_color, false);
 
   /* Top.  */
-  pgtk_fill_rectangle (s->f, s->xgcv.foreground,
-		       left_x, top_y, right_x - left_x + 1, hwidth,
-		       false);
+  cairo_rectangle (cr, left_x, top_y, right_x - left_x + 1, hwidth);
 
   /* Left.  */
   if (left_p)
-    pgtk_fill_rectangle (s->f, s->xgcv.foreground,
-			 left_x, top_y, vwidth, bottom_y - top_y + 1,
-			 false);
+    cairo_rectangle (cr, left_x, top_y, vwidth, bottom_y - top_y + 1);
 
   /* Bottom.  */
-  pgtk_fill_rectangle (s->f, s->xgcv.foreground,
-		       left_x, bottom_y - hwidth + 1, right_x - left_x + 1,
-		       hwidth, false);
+  cairo_rectangle (cr, left_x, bottom_y - hwidth + 1,
+		   right_x - left_x + 1, hwidth);
 
   /* Right.  */
   if (right_p)
-    pgtk_fill_rectangle (s->f, s->xgcv.foreground,
-			 right_x - vwidth + 1, top_y, vwidth,
-			 bottom_y - top_y + 1, false);
-
-  s->xgcv.foreground = foreground_backup;
+    cairo_rectangle (cr, right_x - vwidth + 1, top_y, vwidth,
+		     bottom_y - top_y + 1);
 
+  cairo_fill (cr);
   pgtk_end_cr_clip (s->f);
 }
 
@@ -2001,14 +2047,14 @@ pgtk_draw_glyph_string_box (struct glyph_string *s)
 }
 
 static void
-pgtk_draw_horizontal_wave (struct frame *f, unsigned long color, int x, int y,
-			   int width, int height, int wave_length)
+pgtk_draw_horizontal_wave (struct frame *f, cairo_t *cr, unsigned long color,
+			   int x, int y, int width, int height,
+			   int wave_length)
 {
-  cairo_t *cr;
   double dx = wave_length, dy = height - 1;
   int xoffset, n;
 
-  cr = pgtk_begin_cr_clip (f);
+  cairo_save (cr);
   pgtk_set_cr_source_with_color (f, color, false);
   cairo_rectangle (cr, x, y, width, height);
   cairo_clip (cr);
@@ -2038,16 +2084,17 @@ pgtk_draw_horizontal_wave (struct frame *f, unsigned long color, int x, int y,
     }
   cairo_set_line_width (cr, 1);
   cairo_stroke (cr);
-  pgtk_end_cr_clip (f);
+  cairo_restore (cr);
 }
 
 static void
-pgtk_draw_underwave (struct glyph_string *s, unsigned long color)
+pgtk_draw_underwave (struct glyph_string *s, cairo_t *cr, unsigned long color)
 {
   int wave_height = 3, wave_length = 2;
 
-  pgtk_draw_horizontal_wave (s->f, color, s->x, s->ybase - wave_height + 3,
-			     s->width, wave_height, wave_length);
+  pgtk_draw_horizontal_wave (s->f, cr, color, s->x,
+			     s->ybase - wave_height + 3, s->width,
+			     wave_height, wave_length);
 }
 
 /* Draw a relief around the image glyph string S.  */
@@ -2231,9 +2278,12 @@ pgtk_draw_image_foreground (struct glyph_string *s)
 	  if (s->hl == DRAW_CURSOR)
 	    {
 	      int relief = eabs (s->img->relief);
-	      pgtk_draw_rectangle (s->f, s->xgcv.foreground, x - relief,
-				   y - relief, s->slice.width + relief * 2 - 1,
-				   s->slice.height + relief * 2 - 1, false);
+	      pgtk_set_cr_source_with_color (s->f, s->xgcv.foreground, false);
+	      cairo_rectangle (cr, x - relief + 0.5, y - relief + 0.5,
+			       s->slice.width + relief * 2 - 1,
+			       s->slice.height + relief * 2 - 1);
+	      cairo_set_line_width (cr, 1);
+	      cairo_stroke (cr);
 	    }
 	}
       pgtk_end_cr_clip (s->f);
@@ -2413,14 +2463,25 @@ pgtk_draw_stretch_glyph_string (struct glyph_string *s)
    S, with each segment SEGMENT pixels in length.  */
 
 static void
-pgtk_draw_dash (struct frame *f, struct glyph_string *s,
-		unsigned long foreground, int width,
-		char segment, int offset, int thickness)
+pgtk_fill_decoration_rectangle (struct frame *f, cairo_t *cr,
+				unsigned long color, int x, int y,
+				int width, int height)
+{
+  cairo_save (cr);
+  pgtk_set_cr_source_with_color (f, color, false);
+  cairo_rectangle (cr, x, y, width, height);
+  cairo_fill (cr);
+  cairo_restore (cr);
+}
+
+static void
+pgtk_draw_dash (struct frame *f, struct glyph_string *s, cairo_t *cr,
+		unsigned long foreground, int width, char segment,
+		int offset, int thickness)
 {
-  cairo_t *cr;
   double cr_segment, y_center;
 
-  cr = pgtk_begin_cr_clip (s->f);
+  cairo_save (cr);
   pgtk_set_cr_source_with_color (f, foreground, false);
   cr_segment = (double) segment;
   y_center = s->ybase + offset + (thickness / 2.0);
@@ -2430,7 +2491,7 @@ pgtk_draw_dash (struct frame *f, struct glyph_string *s,
   cairo_move_to (cr, s->x, y_center);
   cairo_line_to (cr, s->x + width, y_center);
   cairo_stroke (cr);
-  pgtk_end_cr_clip (f);
+  cairo_restore (cr);
 }
 
 /* Draw an underline of STYLE onto F at an offset of POSITION from the
@@ -2438,7 +2499,7 @@ pgtk_draw_dash (struct frame *f, struct glyph_string *s,
    DECORATION_WIDTH in length, and THICKNESS in height.  */
 
 static void
-pgtk_fill_underline (struct frame *f, struct glyph_string *s,
+pgtk_fill_underline (struct frame *f, struct glyph_string *s, cairo_t *cr,
 		     unsigned long foreground,
 		     enum face_underline_type style, int position,
 		     int decoration_width, int thickness)
@@ -2454,8 +2515,9 @@ pgtk_fill_underline (struct frame *f, struct glyph_string *s,
 	 function.  */
     case FACE_UNDERLINE_SINGLE:
     case FACE_UNDERLINE_DOUBLE_LINE:
-      pgtk_fill_rectangle (f, foreground, s->x, s->ybase + position,
-			   decoration_width, thickness, false);
+      pgtk_fill_decoration_rectangle (f, cr, foreground, s->x,
+				      s->ybase + position, decoration_width,
+				      thickness);
       break;
 
     case FACE_UNDERLINE_DOTS:
@@ -2463,7 +2525,7 @@ pgtk_fill_underline (struct frame *f, struct glyph_string *s,
       FALLTHROUGH;
 
     case FACE_UNDERLINE_DASHES:
-      pgtk_draw_dash (f, s, foreground, decoration_width, segment,
+      pgtk_draw_dash (f, s, cr, foreground, decoration_width, segment,
 		      position, thickness);
       break;
 
@@ -2589,9 +2651,9 @@ pgtk_draw_glyph_string (struct glyph_string *s)
 	  if (s->face->underline == FACE_UNDERLINE_WAVE)
 	    {
 	      if (s->face->underline_defaulted_p)
-		pgtk_draw_underwave (s, s->xgcv.foreground);
+		pgtk_draw_underwave (s, cr, s->xgcv.foreground);
 	      else
-		pgtk_draw_underwave (s, s->face->underline_color);
+		pgtk_draw_underwave (s, cr, s->face->underline_color);
 	    }
 	  else if (s->face->underline >= FACE_UNDERLINE_SINGLE)
 	    {
@@ -2662,8 +2724,9 @@ pgtk_draw_glyph_string (struct glyph_string *s)
 	      else
 		foreground = s->face->underline_color;
 
-	      pgtk_fill_underline (s->f, s, foreground, s->face->underline,
-				   position, s->width, thickness);
+	      pgtk_fill_underline (s->f, s, cr, foreground,
+				   s->face->underline, position, s->width,
+				   thickness);
 
 	      /* Place a second underline above the first if this was
 		 requested in the face specification.  */
@@ -2672,8 +2735,9 @@ pgtk_draw_glyph_string (struct glyph_string *s)
 		{
 		  /* Compute the position of the second underline.  */
 		  position = position - thickness - 1;
-		  pgtk_fill_underline (s->f, s, foreground, s->face->underline,
-				       position, s->width, thickness);
+		  pgtk_fill_underline (s->f, s, cr, foreground,
+				       s->face->underline, position,
+				       s->width, thickness);
 		}
 	    }
 	}
@@ -2683,11 +2747,11 @@ pgtk_draw_glyph_string (struct glyph_string *s)
 	  unsigned long dy = 0, h = 1;
 
 	  if (s->face->overline_color_defaulted_p)
-	    pgtk_fill_rectangle (s->f, s->xgcv.foreground, s->x, s->y + dy,
-				 s->width, h, false);
+	    pgtk_fill_decoration_rectangle (s->f, cr, s->xgcv.foreground,
+					    s->x, s->y + dy, s->width, h);
 	  else
-	    pgtk_fill_rectangle (s->f, s->face->overline_color, s->x,
-				 s->y + dy, s->width, h, false);
+	    pgtk_fill_decoration_rectangle (s->f, cr, s->face->overline_color,
+					    s->x, s->y + dy, s->width, h);
 	}
 
       /* Draw strike-through.  */
@@ -2706,11 +2770,12 @@ pgtk_draw_glyph_string (struct glyph_string *s)
           unsigned long dy = (glyph_height - h) / 2;
 
 	  if (s->face->strike_through_color_defaulted_p)
-	    pgtk_fill_rectangle (s->f, s->xgcv.foreground, s->x, glyph_y + dy,
-				 s->width, h, false);
+	    pgtk_fill_decoration_rectangle (s->f, cr, s->xgcv.foreground,
+					    s->x, glyph_y + dy, s->width, h);
 	  else
-	    pgtk_fill_rectangle (s->f, s->face->strike_through_color, s->x,
-				 glyph_y + dy, s->width, h, false);
+	    pgtk_fill_decoration_rectangle (s->f, cr,
+					    s->face->strike_through_color,
+					    s->x, glyph_y + dy, s->width, h);
 	}
 
       if (s->prev)
@@ -2861,8 +2926,9 @@ pgtk_draw_hollow_cursor (struct window *w, struct glyph_row *row)
     }
   /* Set clipping, draw the rectangle, and reset clipping again.  */
   pgtk_clip_to_row (w, row, TEXT_AREA, cr);
-  pgtk_draw_rectangle (f, FRAME_X_OUTPUT (f)->cursor_color,
-		       x, y, wd, h - 1, false);
+  cairo_rectangle (cr, x + 0.5, y + 0.5, wd, h - 1);
+  cairo_set_line_width (cr, 1);
+  cairo_stroke (cr);
   pgtk_end_cr_clip (f);
 }
 
@@ -2918,6 +2984,7 @@ pgtk_draw_bar_cursor (struct window *w, struct glyph_row *row, int width,
 	color = FRAME_X_OUTPUT (f)->cursor_color;
 
       pgtk_clip_to_row (w, row, TEXT_AREA, cr);
+      pgtk_set_cr_source_with_color (f, color, false);
 
       if (kind == BAR_CURSOR)
 	{
@@ -2934,9 +3001,9 @@ pgtk_draw_bar_cursor (struct window *w, struct glyph_row *row, int width,
 	  if ((cursor_glyph->resolved_level & 1) != 0)
 	    x += cursor_glyph->pixel_width - width;
 
-	  pgtk_fill_rectangle (f, color, x,
-			       WINDOW_TO_FRAME_PIXEL_Y (w, w->phys_cursor.y),
-			       width, row->height, false);
+	  cairo_rectangle (cr, x,
+			   WINDOW_TO_FRAME_PIXEL_Y (w, w->phys_cursor.y),
+			   width, row->height);
 	}
       else			/* HBAR_CURSOR */
 	{
@@ -2954,12 +3021,13 @@ pgtk_draw_bar_cursor (struct window *w, struct glyph_row *row, int width,
 	  if ((cursor_glyph->resolved_level & 1) != 0
 	      && cursor_glyph->pixel_width > w->phys_cursor_width - 1)
 	    x += cursor_glyph->pixel_width - w->phys_cursor_width + 1;
-	  pgtk_fill_rectangle (f, color, x,
-			       WINDOW_TO_FRAME_PIXEL_Y (w, w->phys_cursor.y +
-							row->height - width),
-			       w->phys_cursor_width - 1, width, false);
+	  cairo_rectangle (cr, x,
+			   WINDOW_TO_FRAME_PIXEL_Y (w, w->phys_cursor.y
+						    + row->height - width),
+			   w->phys_cursor_width - 1, width);
 	}
 
+      cairo_fill (cr);
       pgtk_end_cr_clip (f);
     }
 }
@@ -3036,6 +3104,10 @@ pgtk_copy_bits (struct frame *f, cairo_rectangle_t *src_rect,
   cairo_t *cr;
   cairo_surface_t *surface;	/* temporary surface */
 
+  if (src_rect->width <= 0 || src_rect->height <= 0
+      || dst_rect->width <= 0 || dst_rect->height <= 0)
+    return;
+
   surface
     = cairo_surface_create_similar (FRAME_CR_SURFACE (f),
 				    CAIRO_CONTENT_COLOR_ALPHA,
@@ -3439,7 +3511,9 @@ pgtk_mouse_position (struct frame **fp, int insist, Lisp_Object * bar_window,
   GdkDevice *device;
   GdkModifierType mask;
   GdkWindow *win;
+  GdkWindow *frame_win = NULL;
   bool return_frame_flag = false;
+  bool have_position = false;
 
   block_input ();
 
@@ -3460,13 +3534,18 @@ pgtk_mouse_position (struct frame **fp, int insist, Lisp_Object * bar_window,
   else
     {
       f1 = *fp;
-      win = gtk_widget_get_window (FRAME_GTK_WIDGET (*fp));
+      frame_win = gtk_widget_get_window (FRAME_GTK_WIDGET (*fp));
       seat = gdk_display_get_default_seat (dpyinfo->gdpy);
       device = gdk_seat_get_pointer (seat);
-      win = gdk_window_get_device_position (win, device, &win_x,
+      win = gdk_window_get_device_position (frame_win, device, &win_x,
 					    &win_y, &mask);
       if (win != NULL)
-	f1 = pgtk_any_window_to_frame (win);
+	{
+	  if (win == frame_win)
+	    have_position = true;
+	  else
+	    f1 = pgtk_any_window_to_frame (win);
+	}
       else
 	{
 	  f1 = SELECTED_FRAME ();
@@ -3485,12 +3564,15 @@ pgtk_mouse_position (struct frame **fp, int insist, Lisp_Object * bar_window,
       return;
     }
 
-  win = gtk_widget_get_window (FRAME_GTK_WIDGET (f1));
-  seat = gdk_display_get_default_seat (dpyinfo->gdpy);
-  device = gdk_seat_get_pointer (seat);
+  if (!have_position)
+    {
+      win = gtk_widget_get_window (FRAME_GTK_WIDGET (f1));
+      seat = gdk_display_get_default_seat (dpyinfo->gdpy);
+      device = gdk_seat_get_pointer (seat);
 
-  win = gdk_window_get_device_position (win, device,
-					&win_x, &win_y, &mask);
+      win = gdk_window_get_device_position (win, device,
+					    &win_x, &win_y, &mask);
+    }
 
   if (f1 != NULL)
     {
@@ -3600,7 +3682,7 @@ pgtk_draw_fringe_bitmap (struct window *w, struct glyph_row *row,
   pgtk_clip_to_row (w, row, ANY_AREA, cr);
 
   if (p->bx >= 0 && !p->overlay_p)
-    fill_background_by_face (f, face, p->bx, p->by, p->nx, p->ny);
+    fill_background_by_face_1 (f, face, cr, p->bx, p->by, p->nx, p->ny);
 
   if (p->which
       && p->which < max_fringe_bmp
@@ -4708,8 +4790,8 @@ frame_unhighlight (struct frame *f)
 }
 
 
-void
-pgtk_frame_rehighlight (struct pgtk_display_info *dpyinfo)
+static void
+pgtk_frame_rehighlight_1 (struct pgtk_display_info *dpyinfo, bool force)
 {
   struct frame *old_highlight = dpyinfo->highlight_frame;
 
@@ -4728,12 +4810,27 @@ pgtk_frame_rehighlight (struct pgtk_display_info *dpyinfo)
   else
     dpyinfo->highlight_frame = 0;
 
+  if (!force && old_highlight == dpyinfo->highlight_frame)
+    return;
+
   if (old_highlight)
     frame_unhighlight (old_highlight);
   if (dpyinfo->highlight_frame)
     frame_highlight (dpyinfo->highlight_frame);
 }
 
+void
+pgtk_frame_rehighlight (struct pgtk_display_info *dpyinfo)
+{
+  pgtk_frame_rehighlight_1 (dpyinfo, false);
+}
+
+void
+pgtk_frame_rehighlight_forced (struct pgtk_display_info *dpyinfo)
+{
+  pgtk_frame_rehighlight_1 (dpyinfo, true);
+}
+
 /* The focus has changed, or we have redirected a frame's focus to
    another frame (this happens when a frame uses a surrogate
    mini-buffer frame).  Shift the highlight as appropriate.
@@ -4932,9 +5029,9 @@ pgtk_any_window_to_frame (GdkWindow *window)
 }
 
 static gboolean
-pgtk_handle_event (GtkWidget *widget, GdkEvent *event, gpointer *data)
+pgtk_handle_event (GtkWidget *widget, GdkEvent *event, gpointer data)
 {
-  struct frame *f;
+  struct frame *f = data;
   union buffered_input_event inev;
   GtkWidget *frame_widget;
   gint x, y;
@@ -4943,30 +5040,32 @@ pgtk_handle_event (GtkWidget *widget, GdkEvent *event, gpointer *data)
       && (event->touchpad_pinch.phase
 	  != GDK_TOUCHPAD_GESTURE_PHASE_END))
     {
-      f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+      if (f == NULL)
+	f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+      if (f == NULL)
+	return FALSE;
       frame_widget = FRAME_GTK_WIDGET (f);
 
       gtk_widget_translate_coordinates (widget, frame_widget,
 					lrint (event->touchpad_pinch.x),
 					lrint (event->touchpad_pinch.y),
 					&x, &y);
-      if (f)
-	{
 
-	  inev.ie.kind = PINCH_EVENT;
-	  XSETFRAME (inev.ie.frame_or_window, f);
-	  XSETINT (inev.ie.x, x);
-	  XSETINT (inev.ie.y, y);
-	  inev.ie.arg = list4 (make_float (event->touchpad_pinch.dx),
-			       make_float (event->touchpad_pinch.dy),
-			       make_float (event->touchpad_pinch.scale),
-			       make_float (event->touchpad_pinch.angle_delta));
-	  inev.ie.modifiers = pgtk_gtk_to_emacs_modifiers (FRAME_DISPLAY_INFO (f),
-							   event->touchpad_pinch.state);
-	  inev.ie.device
-	    = pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f), event);
-	  evq_enqueue (&inev);
-	}
+      struct pgtk_display_info *dpyinfo = FRAME_DISPLAY_INFO (f);
+
+      inev.ie.kind = PINCH_EVENT;
+      XSETFRAME (inev.ie.frame_or_window, f);
+      XSETINT (inev.ie.x, x);
+      XSETINT (inev.ie.y, y);
+      inev.ie.arg = list4 (make_float (event->touchpad_pinch.dx),
+			   make_float (event->touchpad_pinch.dy),
+			   make_float (event->touchpad_pinch.scale),
+			   make_float (event->touchpad_pinch.angle_delta));
+      inev.ie.modifiers = pgtk_gtk_to_emacs_modifiers (dpyinfo,
+						       event->touchpad_pinch.state);
+      inev.ie.device
+	= pgtk_get_device_for_event (dpyinfo, event);
+      evq_enqueue (&inev);
 
       return TRUE;
     }
@@ -5012,14 +5111,18 @@ pgtk_clear_under_internal_border (struct frame *f)
 
       if (face)
 	{
-	  fill_background_by_face (f, face, 0, margin, width, border);
-	  fill_background_by_face (f, face, 0, 0, border, height);
-	  fill_background_by_face (f, face, width - border, 0, border,
-				   height);
-	  fill_background_by_face (f, face, 0, (height
-						- bottom_margin
-						- border),
-				   width, border);
+	  cairo_t *cr = pgtk_begin_cr_clip (f);
+
+	  fill_background_by_face_1 (f, face, cr, 0, margin, width, border);
+	  fill_background_by_face_1 (f, face, cr, 0, 0, border, height);
+	  fill_background_by_face_1 (f, face, cr, width - border, 0, border,
+				     height);
+	  fill_background_by_face_1 (f, face, cr, 0, (height
+						  - bottom_margin
+						  - border),
+				     width, border);
+
+	  pgtk_end_cr_clip (f);
 	}
       else
 	{
@@ -5035,16 +5138,17 @@ pgtk_clear_under_internal_border (struct frame *f)
 }
 
 static gboolean
-pgtk_handle_draw (GtkWidget *widget, cairo_t *cr, gpointer *data)
+pgtk_handle_draw (GtkWidget *widget, cairo_t *cr, gpointer data)
 {
-  struct frame *f;
+  struct frame *f = data;
 
   GdkWindow *win = gtk_widget_get_window (widget);
 
   if (win != NULL)
     {
       cairo_surface_t *src = NULL;
-      f = pgtk_any_window_to_frame (win);
+      if (f == NULL)
+	f = pgtk_any_window_to_frame (win);
       if (f != NULL)
 	{
 	  src = FRAME_X_OUTPUT (f)->cr_surface_visible_bell;
@@ -5064,10 +5168,10 @@ pgtk_handle_draw (GtkWidget *widget, cairo_t *cr, gpointer *data)
 size_allocate (GtkWidget *widget, GtkAllocation *alloc,
 	       gpointer user_data)
 {
-  struct frame *f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+  struct frame *f = user_data;
 
   if (!f)
-    f = user_data;
+    f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
 
   if (f)
     {
@@ -5213,15 +5317,14 @@ pgtk_enqueue_preedit (struct frame *f, Lisp_Object preedit)
 }
 
 static gboolean
-key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
+key_press_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
   union buffered_input_event inev;
   ptrdiff_t nbytes;
   Mouse_HLInfo *hlinfo;
-  struct frame *f;
+  struct frame *f = (struct frame *) user_data;
   struct pgtk_display_info *dpyinfo;
 
-  f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
   EVENT_INIT (inev.ie);
   hlinfo = MOUSE_HL_INFO (f);
   nbytes = 0;
@@ -5272,7 +5375,7 @@ key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
 	    return TRUE;
 	}
 
-      state |= pgtk_emacs_to_gtk_modifiers (FRAME_DISPLAY_INFO (f),
+      state |= pgtk_emacs_to_gtk_modifiers (dpyinfo,
 					    extra_keyboard_modifiers);
       modifiers = state;
 
@@ -5295,7 +5398,7 @@ key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
       /* Common for all keysym input events.  */
       XSETFRAME (inev.ie.frame_or_window, f);
       inev.ie.modifiers
-	= pgtk_gtk_to_emacs_modifiers (FRAME_DISPLAY_INFO (f), modifiers);
+	= pgtk_gtk_to_emacs_modifiers (dpyinfo, modifiers);
       inev.ie.timestamp = event->key.time;
 
       /* First deal with keysyms which have defined
@@ -5307,7 +5410,7 @@ key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
 	  inev.ie.code = keysym;
 
 	  inev.ie.device
-	    = pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f), event);
+	    = pgtk_get_device_for_event (dpyinfo, event);
 	  goto done;
 	}
 
@@ -5321,7 +5424,7 @@ key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
 	  inev.ie.code = keysym & 0xFFFFFF;
 
 	  inev.ie.device
-	    = pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f), event);
+	    = pgtk_get_device_for_event (dpyinfo, event);
 	  goto done;
 	}
 
@@ -5336,7 +5439,7 @@ key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
 	  inev.ie.code = XFIXNAT (c);
 
 	  inev.ie.device
-	    = pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f), event);
+	    = pgtk_get_device_for_event (dpyinfo, event);
 	  goto done;
 	}
 
@@ -5418,7 +5521,7 @@ key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
 	  inev.ie.code = keysym;
 
 	  inev.ie.device
-	    = pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f), event);
+	    = pgtk_get_device_for_event (dpyinfo, event);
 	  goto done;
 	}
 
@@ -5426,7 +5529,7 @@ key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
 	inev.ie.kind = MULTIBYTE_CHAR_KEYSTROKE_EVENT;
 	inev.ie.arg = make_unibyte_string ((char *) copy_bufptr, nbytes);
 	inev.ie.device
-	  = pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f), event);
+	  = pgtk_get_device_for_event (dpyinfo, event);
 
 	if (keysym == GDK_KEY_VoidSymbol)
 	  goto done;
@@ -5443,30 +5546,11 @@ key_press_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
   return TRUE;
 }
 
-static struct pgtk_display_info *
-pgtk_display_info_for_display (GdkDisplay *dpy)
-{
-  struct pgtk_display_info *dpyinfo;
-
-  for (dpyinfo = x_display_list; dpyinfo; dpyinfo = dpyinfo->next)
-    {
-      if (dpyinfo->display == dpy)
-	return dpyinfo;
-    }
-
-  return NULL;
-}
-
 static gboolean
-key_release_event (GtkWidget *widget,
-		   GdkEvent *event,
-		   gpointer *user_data)
+key_release_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
-  GdkDisplay *display;
-  struct pgtk_display_info *dpyinfo;
-
-  display = gtk_widget_get_display (widget);
-  dpyinfo = pgtk_display_info_for_display (display);
+  struct frame *f = (struct frame *) user_data;
+  struct pgtk_display_info *dpyinfo = FRAME_DISPLAY_INFO (f);
 
   if (dpyinfo)
     /* This is needed on Wayland because of some brain dead
@@ -5480,9 +5564,9 @@ key_release_event (GtkWidget *widget,
 static gboolean
 configure_event (GtkWidget *widget,
 		 GdkEvent *event,
-		 gpointer *user_data)
+		 gpointer user_data)
 {
-  struct frame *f = pgtk_any_window_to_frame (event->configure.window);
+  struct frame *f = user_data;
 
   if (f && widget == FRAME_GTK_OUTER_WIDGET (f))
     {
@@ -5512,9 +5596,9 @@ configure_event (GtkWidget *widget,
 static gboolean
 map_event (GtkWidget *widget,
 	   GdkEvent *event,
-	   gpointer *user_data)
+	   gpointer user_data)
 {
-  struct frame *f = pgtk_any_window_to_frame (event->any.window);
+  struct frame *f = user_data;
   union buffered_input_event inev;
 
   EVENT_INIT (inev.ie);
@@ -5559,9 +5643,9 @@ map_event (GtkWidget *widget,
 static gboolean
 window_state_event (GtkWidget *widget,
 		    GdkEvent *event,
-		    gpointer *user_data)
+		    gpointer user_data)
 {
-  struct frame *f = pgtk_any_window_to_frame (event->window_state.window);
+  struct frame *f = user_data;
   GdkWindowState new_state;
   union buffered_input_event inev;
 
@@ -5633,9 +5717,9 @@ window_state_event (GtkWidget *widget,
 
 static gboolean
 delete_event (GtkWidget *widget,
-	      GdkEvent *event, gpointer *user_data)
+	      GdkEvent *event, gpointer user_data)
 {
-  struct frame *f = pgtk_any_window_to_frame (event->any.window);
+  struct frame *f = user_data;
   union buffered_input_event inev;
 
   EVENT_INIT (inev.ie);
@@ -5709,12 +5793,10 @@ pgtk_focus_changed (gboolean is_enter, int state,
 }
 
 static gboolean
-enter_notify_event (GtkWidget *widget, GdkEvent *event,
-		    gpointer *user_data)
+enter_notify_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
   union buffered_input_event inev;
-  struct frame *frame
-    = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+  struct frame *frame = (struct frame *) user_data;
 
   if (frame == NULL)
     return FALSE;
@@ -5737,12 +5819,10 @@ enter_notify_event (GtkWidget *widget, GdkEvent *event,
 }
 
 static gboolean
-leave_notify_event (GtkWidget *widget, GdkEvent *event,
-		    gpointer *user_data)
+leave_notify_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
   union buffered_input_event inev;
-  struct frame *frame
-    = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+  struct frame *frame = (struct frame *) user_data;
 
   if (frame == NULL)
     return FALSE;
@@ -5786,11 +5866,10 @@ leave_notify_event (GtkWidget *widget, GdkEvent *event,
 }
 
 static gboolean
-focus_in_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
+focus_in_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
   union buffered_input_event inev;
-  struct frame *frame
-    = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+  struct frame *frame = (struct frame *) user_data;
 
   if (frame == NULL)
     return TRUE;
@@ -5810,11 +5889,10 @@ focus_in_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
 }
 
 static gboolean
-focus_out_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
+focus_out_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
   union buffered_input_event inev;
-  struct frame *frame
-    = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+  struct frame *frame = (struct frame *) user_data;
 
   if (frame == NULL)
     return TRUE;
@@ -5864,8 +5942,7 @@ note_mouse_movement (struct frame *frame,
       note_mouse_highlight (frame, -1, -1);
       dpyinfo->last_mouse_glyph_frame = NULL;
       frame->last_mouse_device
-	= pgtk_get_device_for_event (FRAME_DISPLAY_INFO (frame),
-				     (GdkEvent *) event);
+	= pgtk_get_device_for_event (dpyinfo, (GdkEvent *) event);
       return true;
     }
 
@@ -5883,8 +5960,7 @@ note_mouse_movement (struct frame *frame,
       remember_mouse_glyph (frame, event->x, event->y, r);
       dpyinfo->last_mouse_glyph_frame = frame;
       frame->last_mouse_device
-	= pgtk_get_device_for_event (FRAME_DISPLAY_INFO (frame),
-				     (GdkEvent *) event);
+	= pgtk_get_device_for_event (dpyinfo, (GdkEvent *) event);
       return true;
     }
 
@@ -5892,11 +5968,10 @@ note_mouse_movement (struct frame *frame,
 }
 
 static gboolean
-motion_notify_event (GtkWidget *widget, GdkEvent *event,
-		     gpointer *user_data)
+motion_notify_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
   union buffered_input_event inev;
-  struct frame *f, *frame;
+  struct frame *f, *frame = (struct frame *) user_data;
   struct pgtk_display_info *dpyinfo;
   Mouse_HLInfo *hlinfo;
   GdkDevice *device;
@@ -5918,10 +5993,8 @@ motion_notify_event (GtkWidget *widget, GdkEvent *event,
   previous_help_echo_string = help_echo_string;
   help_echo_string = Qnil;
 
-  frame = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
   dpyinfo = FRAME_DISPLAY_INFO (frame);
-  f = (gui_mouse_grabbed (dpyinfo) ? dpyinfo->last_mouse_frame
-       : pgtk_any_window_to_frame (gtk_widget_get_window (widget)));
+  f = gui_mouse_grabbed (dpyinfo) ? dpyinfo->last_mouse_frame : frame;
   hlinfo = MOUSE_HL_INFO (f);
 
   if (hlinfo->mouse_face_hidden)
@@ -6016,12 +6089,14 @@ construct_mouse_click (struct input_event *result,
 		       const GdkEventButton *event,
 		       struct frame *f)
 {
+  struct pgtk_display_info *dpyinfo = FRAME_DISPLAY_INFO (f);
+
   /* Make the event type NO_EVENT; we'll change that when we decide
      otherwise.  */
   result->kind = MOUSE_CLICK_EVENT;
   result->code = event->button - 1;
   result->timestamp = event->time;
-  result->modifiers = (pgtk_gtk_to_emacs_modifiers (FRAME_DISPLAY_INFO (f),
+  result->modifiers = (pgtk_gtk_to_emacs_modifiers (dpyinfo,
 						    event->state)
 		       | (event->type == GDK_BUTTON_RELEASE
 			  ? up_modifier : down_modifier));
@@ -6030,17 +6105,15 @@ construct_mouse_click (struct input_event *result,
   XSETINT (result->y, event->y);
   XSETFRAME (result->frame_or_window, f);
   result->arg = Qnil;
-  result->device = pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f),
-					      (GdkEvent *) event);
+  result->device = pgtk_get_device_for_event (dpyinfo, (GdkEvent *) event);
   return Qnil;
 }
 
 static gboolean
-button_event (GtkWidget *widget, GdkEvent *event,
-	      gpointer *user_data)
+button_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
   union buffered_input_event inev;
-  struct frame *f, *frame;
+  struct frame *f, *frame = (struct frame *) user_data;
   struct pgtk_display_info *dpyinfo;
 
   /* If we decide we want to generate an event to be seen
@@ -6068,7 +6141,6 @@ button_event (GtkWidget *widget, GdkEvent *event,
   if (event->type != GDK_BUTTON_PRESS && event->type != GDK_BUTTON_RELEASE)
     return TRUE;
 
-  frame = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
   dpyinfo = FRAME_DISPLAY_INFO (frame);
 
   dpyinfo->last_mouse_glyph_frame = NULL;
@@ -6077,7 +6149,7 @@ button_event (GtkWidget *widget, GdkEvent *event,
     f = dpyinfo->last_mouse_frame;
   else
     {
-      f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+      f = frame;
 
       if (f && event->button.type == GDK_BUTTON_PRESS
 	  && !FRAME_NO_ACCEPT_FOCUS (f))
@@ -6168,11 +6240,11 @@ button_event (GtkWidget *widget, GdkEvent *event,
 }
 
 static gboolean
-scroll_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
+scroll_event (GtkWidget *widget, GdkEvent *event, gpointer user_data)
 {
   union buffered_input_event inev;
-  struct frame *f, *frame;
-  struct pgtk_display_info *dpyinfo;
+  struct frame *f, *frame = (struct frame *) user_data;
+  struct pgtk_display_info *dpyinfo, *f_dpyinfo;
   GdkScrollDirection dir;
   double delta_x, delta_y;
 
@@ -6180,18 +6252,15 @@ scroll_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
   inev.ie.kind = NO_EVENT;
   inev.ie.arg = Qnil;
 
-  frame = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
   dpyinfo = FRAME_DISPLAY_INFO (frame);
 
-  if (gui_mouse_grabbed (dpyinfo))
-    f = dpyinfo->last_mouse_frame;
-  else
-    f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
+  f = gui_mouse_grabbed (dpyinfo) ? dpyinfo->last_mouse_frame : frame;
+  f_dpyinfo = FRAME_DISPLAY_INFO (f);
 
   inev.ie.kind = NO_EVENT;
   inev.ie.timestamp = event->scroll.time;
   inev.ie.modifiers
-    = pgtk_gtk_to_emacs_modifiers (FRAME_DISPLAY_INFO (f), event->scroll.state);
+    = pgtk_gtk_to_emacs_modifiers (f_dpyinfo, event->scroll.state);
   XSETINT (inev.ie.x, event->scroll.x);
   XSETINT (inev.ie.y, event->scroll.y);
   XSETFRAME (inev.ie.frame_or_window, f);
@@ -6201,7 +6270,7 @@ scroll_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
     {
       inev.ie.kind = TOUCH_END_EVENT;
       inev.ie.device
-	= pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f), event);
+	= pgtk_get_device_for_event (f_dpyinfo, event);
       evq_enqueue (&inev);
       return TRUE;
     }
@@ -6297,7 +6366,7 @@ scroll_event (GtkWidget *widget, GdkEvent *event, gpointer *user_data)
   if (inev.ie.kind != NO_EVENT)
     {
       inev.ie.device
-	= pgtk_get_device_for_event (FRAME_DISPLAY_INFO (f), event);
+	= pgtk_get_device_for_event (f_dpyinfo, event);
       evq_enqueue (&inev);
     }
   return TRUE;
@@ -6390,11 +6459,9 @@ pgtk_finish_drop (Lisp_Object success, Lisp_Object event_time,
 drag_leave (GtkWidget *widget, GdkDragContext *context,
 	    guint time, gpointer user_data)
 {
-  struct frame *f;
+  struct frame *f = user_data;
   union buffered_input_event inev;
 
-  f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
-
   if (current_drop_context)
     {
       if (current_drop_context_drop)
@@ -6421,16 +6488,14 @@ drag_leave (GtkWidget *widget, GdkDragContext *context,
 
 static gboolean
 drag_motion (GtkWidget *widget, GdkDragContext *context,
-             gint x, gint y, guint time)
-
+	     gint x, gint y, guint time, gpointer user_data)
 {
-  struct frame *f;
+  struct frame *f = user_data;
   union buffered_input_event inev;
   GdkAtom name;
+  gchar *name_string;
   GdkDragAction suggestion;
 
-  f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
-
   if (!f)
     return FALSE;
 
@@ -6452,12 +6517,14 @@ drag_motion (GtkWidget *widget, GdkDragContext *context,
   suggestion = gdk_drag_context_get_suggested_action (context);
 
   EVENT_INIT (inev.ie);
+  name_string = gdk_atom_name (name);
 
   inev.ie.kind = DRAG_N_DROP_EVENT;
   inev.ie.modifiers = 0;
-  inev.ie.arg = list4 (Qlambda, intern (gdk_atom_name (name)),
+  inev.ie.arg = list4 (Qlambda, intern (name_string),
 		       make_uint (time),
 		       drag_action_to_symbol (suggestion));
+  g_free (name_string);
   inev.ie.timestamp = time;
 
   XSETINT (inev.ie.x, x);
@@ -6473,13 +6540,12 @@ drag_motion (GtkWidget *widget, GdkDragContext *context,
 drag_drop (GtkWidget *widget, GdkDragContext *context,
 	   int x, int y, guint time, gpointer user_data)
 {
-  struct frame *f;
+  struct frame *f = user_data;
   union buffered_input_event inev;
   GdkAtom name;
+  gchar *name_string;
   GdkDragAction selected_action;
 
-  f = pgtk_any_window_to_frame (gtk_widget_get_window (widget));
-
   if (!f)
     return FALSE;
 
@@ -6501,12 +6567,14 @@ drag_drop (GtkWidget *widget, GdkDragContext *context,
   selected_action = gdk_drag_context_get_selected_action (context);
 
   EVENT_INIT (inev.ie);
+  name_string = gdk_atom_name (name);
 
   inev.ie.kind = DRAG_N_DROP_EVENT;
   inev.ie.modifiers = 0;
-  inev.ie.arg = list4 (Qquote, intern (gdk_atom_name (name)),
+  inev.ie.arg = list4 (Qquote, intern (name_string),
 		       make_uint (time),
 		       drag_action_to_symbol (selected_action));
+  g_free (name_string);
   inev.ie.timestamp = time;
 
   XSETINT (inev.ie.x, x);
@@ -6643,10 +6711,12 @@ touch_event_cb (GtkWidget *self, GdkEvent *event, gpointer user_data)
   struct pgtk_touch_point *touchpoint;
   Lisp_Object arg = Qnil;
   int state;
+  int x = lrint (event->touch.x);
+  int y = lrint (event->touch.y);
 
   EVENT_INIT (inev.ie);
 
-  f = pgtk_any_window_to_frame (gtk_widget_get_window (self));
+  f = user_data;
   eassert (f);
   dpyinfo = FRAME_DISPLAY_INFO (f);
   switch (event->type)
@@ -6666,8 +6736,8 @@ touch_event_cb (GtkWidget *self, GdkEvent *event, gpointer user_data)
       inev.ie.kind = TOUCHSCREEN_BEGIN_EVENT;
       inev.ie.timestamp = event->touch.time;
       XSETFRAME (inev.ie.frame_or_window, f);
-      XSETINT (inev.ie.x, lrint (event->touch.x));
-      XSETINT (inev.ie.y, lrint (event->touch.y));
+      XSETINT (inev.ie.x, x);
+      XSETINT (inev.ie.y, y);
       XSETINT (inev.ie.arg, local_detail);
       break;
 
@@ -6678,13 +6748,12 @@ touch_event_cb (GtkWidget *self, GdkEvent *event, gpointer user_data)
       if (!touchpoint
 	  /* Don't send this event if nothing has changed
 	     either.  */
-	  || (touchpoint->x == lrint (event->touch.x)
-	      && touchpoint->y == lrint (event->touch.y)))
+	  || (touchpoint->x == x && touchpoint->y == y))
 	break;
 
       /* Construct the input event.  */
-      touchpoint->x = lrint (event->touch.x);
-      touchpoint->y = lrint (event->touch.y);
+      touchpoint->x = x;
+      touchpoint->y = y;
       inev.ie.kind = TOUCHSCREEN_UPDATE_EVENT;
       inev.ie.timestamp = event->touch.time;
       XSETFRAME (inev.ie.frame_or_window, f);
@@ -6716,8 +6785,8 @@ touch_event_cb (GtkWidget *self, GdkEvent *event, gpointer user_data)
 	  inev.ie.modifiers = (event->type != GDK_TOUCH_END);
 
 	  XSETFRAME (inev.ie.frame_or_window, f);
-	  XSETINT (inev.ie.x, lrint (event->touch.x));
-	  XSETINT (inev.ie.y, lrint (event->touch.y));
+	  XSETINT (inev.ie.x, x);
+	  XSETINT (inev.ie.y, y);
 	  XSETINT (inev.ie.arg, local_detail);
 	}
       break;
@@ -6775,7 +6844,7 @@ pgtk_set_event_handler (struct frame *f)
   if (f->tooltip)
     {
       g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "draw",
-			G_CALLBACK (pgtk_handle_draw), NULL);
+			G_CALLBACK (pgtk_handle_draw), f);
       return;
     }
 
@@ -6787,61 +6856,61 @@ pgtk_set_event_handler (struct frame *f)
     {
       g_signal_connect (G_OBJECT (FRAME_GTK_OUTER_WIDGET (f)),
 			"window-state-event", G_CALLBACK (window_state_event),
-			NULL);
+			f);
       g_signal_connect (G_OBJECT (FRAME_GTK_OUTER_WIDGET (f)), "delete-event",
-			G_CALLBACK (delete_event), NULL);
+			G_CALLBACK (delete_event), f);
       g_signal_connect (G_OBJECT (FRAME_GTK_OUTER_WIDGET (f)), "event",
-			G_CALLBACK (pgtk_handle_event), NULL);
+			G_CALLBACK (pgtk_handle_event), f);
       g_signal_connect (G_OBJECT (FRAME_GTK_OUTER_WIDGET (f)), "configure-event",
-			G_CALLBACK (configure_event), NULL);
+			G_CALLBACK (configure_event), f);
     }
 
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "map-event",
-		    G_CALLBACK (map_event), NULL);
+		    G_CALLBACK (map_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "size-allocate",
 		    G_CALLBACK (size_allocate), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "key-press-event",
-		    G_CALLBACK (key_press_event), NULL);
+		    G_CALLBACK (key_press_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "key-release-event",
-		    G_CALLBACK (key_release_event), NULL);
+		    G_CALLBACK (key_release_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "focus-in-event",
-		    G_CALLBACK (focus_in_event), NULL);
+		    G_CALLBACK (focus_in_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "focus-out-event",
-		    G_CALLBACK (focus_out_event), NULL);
+		    G_CALLBACK (focus_out_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "enter-notify-event",
-		    G_CALLBACK (enter_notify_event), NULL);
+		    G_CALLBACK (enter_notify_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "leave-notify-event",
-		    G_CALLBACK (leave_notify_event), NULL);
+		    G_CALLBACK (leave_notify_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "motion-notify-event",
-		    G_CALLBACK (motion_notify_event), NULL);
+		    G_CALLBACK (motion_notify_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "button-press-event",
-		    G_CALLBACK (button_event), NULL);
+		    G_CALLBACK (button_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "button-release-event",
-		    G_CALLBACK (button_event), NULL);
+		    G_CALLBACK (button_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "scroll-event",
-		    G_CALLBACK (scroll_event), NULL);
+		    G_CALLBACK (scroll_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "configure-event",
-		    G_CALLBACK (configure_event), NULL);
+		    G_CALLBACK (configure_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "drag-leave",
-		    G_CALLBACK (drag_leave), NULL);
+		    G_CALLBACK (drag_leave), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "drag-motion",
-		    G_CALLBACK (drag_motion), NULL);
+		    G_CALLBACK (drag_motion), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "drag-drop",
-		    G_CALLBACK (drag_drop), NULL);
+		    G_CALLBACK (drag_drop), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "draw",
-		    G_CALLBACK (pgtk_handle_draw), NULL);
+		    G_CALLBACK (pgtk_handle_draw), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "property-notify-event",
-		    G_CALLBACK (pgtk_selection_event), NULL);
+		    G_CALLBACK (pgtk_selection_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "selection-clear-event",
-		    G_CALLBACK (pgtk_selection_event), NULL);
+		    G_CALLBACK (pgtk_selection_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "selection-request-event",
-		    G_CALLBACK (pgtk_selection_event), NULL);
+		    G_CALLBACK (pgtk_selection_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "selection-notify-event",
-		    G_CALLBACK (pgtk_selection_event), NULL);
+		    G_CALLBACK (pgtk_selection_event), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "touch-event",
-		    G_CALLBACK (touch_event_cb), NULL);
+		    G_CALLBACK (touch_event_cb), f);
   g_signal_connect (G_OBJECT (FRAME_GTK_WIDGET (f)), "event",
-		    G_CALLBACK (pgtk_handle_event), NULL);
+		    G_CALLBACK (pgtk_handle_event), f);
 }
 
 static void
@@ -6921,7 +6990,7 @@ pgtk_find_selection_owner (GdkWindow *window)
 pgtk_selection_event (GtkWidget *widget, GdkEvent *event,
 		      gpointer user_data)
 {
-  struct frame *f;
+  struct frame *f = user_data;
   union buffered_input_event inev;
 
   if (event->type == GDK_PROPERTY_NOTIFY)
@@ -6929,7 +6998,8 @@ pgtk_selection_event (GtkWidget *widget, GdkEvent *event,
   else if (event->type == GDK_SELECTION_CLEAR
 	   || event->type == GDK_SELECTION_REQUEST)
     {
-      f = pgtk_find_selection_owner (event->selection.window);
+      if (!f || FRAME_GDK_WINDOW (f) != event->selection.window)
+	f = pgtk_find_selection_owner (event->selection.window);
 
       if (f)
 	{
diff --git a/src/pgtkterm.h b/src/pgtkterm.h
index abf16b3aef1..c54d38eba01 100644
--- a/src/pgtkterm.h
+++ b/src/pgtkterm.h
@@ -258,6 +258,16 @@ #define BLUE_FROM_ULONG(color)	((color) & 0xff)
   /* List of all devices for all seats on this display.  */
   struct pgtk_device_t *devices;
 
+  /* Most recent event device lookup.  */
+  GdkDevice *last_event_device;
+  Lisp_Object last_event_device_name;
+
+  /* Common selection atoms.  */
+  GdkAtom atom_ATOM_PAIR;
+  GdkAtom atom_COMPOUND_TEXT;
+  GdkAtom atom_NULL;
+  GdkAtom atom_UTF8_STRING;
+
   /* The frame where the mouse was last time we reported a mouse position.  */
   struct frame *last_mouse_glyph_frame;
 
@@ -646,6 +656,7 @@ #define SELECTION_EVENT_TIME(eventp)	\
 extern Lisp_Object pgtk_get_focus_frame (struct frame *);
 
 extern void pgtk_frame_rehighlight (struct pgtk_display_info *);
+extern void pgtk_frame_rehighlight_forced (struct pgtk_display_info *);
 
 extern void pgtk_change_tab_bar_height (struct frame *, int);
 
